frontend/src/components/tabs/InputOutputTab.tsx

```diff
@@ -0,0 +1,234 @@
+            />
+          )}
+
+          <div className={commonCss.header}>{'Input'}</div>
+          <DetailsTable
+            key={`input-parameters-${executionId}`}
+            title='Parameters'
+            fields={inputParams}
+          />
+
+          <ArtifactPreview
+            key={`input-artifacts-${executionId}`}
+            title='Artifacts'
+            fields={inputArtifacts}
+          />
+
+          <div className={padding(20, 'tb')} />
+          <hr />
+
+          <div className={commonCss.header}>{'Output'}</div>
+          <DetailsTable
+            key={`output-parameters-${executionId}`}
+            title='Parameters'
+            fields={outputParams}
+          />
+
+          <ArtifactPreview
+            key={`output-artifacts-${executionId}`}
+            title='Artifacts'
+            fields={outputArtifacts}
+          />
+        </div>
+      </div>
+    </ErrorBoundary>
+  );
+}
+
+export default InputOutputTab;
+
+function extractInputFromExecution(execution: Execution): KeyValue<string>[] {
+  return extractParamFromExecution(execution, /input:(?<inputName>.+)/, 'inputName');
+}
+
+function extractOutputFromExecution(execution: Execution): KeyValue<string>[] {
+  return extractParamFromExecution(execution, /output:(?<outputName>.+)/, 'outputName');
+}
+
+function extractParamFromExecution(
+  execution: Execution,
+  pattern: RegExp,
+  groupName: string,
+): KeyValue<string>[] {
+  const result: KeyValue<string>[] = [];
+  execution.getCustomPropertiesMap().forEach((value, key) => {
+    const found = key.match(pattern);
+    if (found?.groups?.[groupName]) {
+      result.push([found.groups[groupName], prettyPrintValue(getMetadataValue(value))]);
+    }
+  });
+  return result;
+}
+
+function prettyPrintValue(value: string | number | Struct | undefined): string {
+  if (value == null) {
+    return '';
+  }
+  if (typeof value === 'string') {
+    return value;
+  }
+  if (typeof value === 'number') {
+    return JSON.stringify(value);
+  }
+  // value is Struct
+  const jsObject = value.toJavaScript();
+  // When Struct is converted to js object, it may contain a top level "struct"
+  // or "list" key depending on its type, but the key is meaningless and we can
+  // omit it in visualization.
+  return JSON.stringify(jsObject?.struct || jsObject?.list || jsObject, null, 2);
+}
+
+function getArtifactParamList(inputArtifactData: Artifact[]): ParamList {
+  return inputArtifactData.map(artifact => [
+    (getResourceProperty(artifact, ArtifactProperties.NAME) ||
+      getResourceProperty(artifact, ArtifactCustomProperties.NAME, true) ||
+      '') as string,
+    artifact.getUri(),
+  ]);
+}
+
+interface ArtifactPreviewProps extends ValueComponentProps<string> {
+  fields: Array<KeyValue<string>>;
+  title?: string;
+}
+
+const ArtifactPreview: React.FC<ArtifactPreviewProps> = (props: ArtifactPreviewProps) => {
+  const { fields, title } = props;
+  return (
+    <>
```

Comment:
Not a blocker here, but I generally feel we should separate html and js files, so we have better readability of the code, and shorten some of the code. i.e. we can make each html section a component and use them here.

Reply:
I assume you got the preference mostly from Angular background (please correct me if that's wrong), note that in React, it embraces a different paradigm.

https://reactjs.org/docs/introducing-jsx.html#why-jsx
> Instead of artificially separating technologies by putting markup and logic in separate files, React separates concerns with loosely coupled units called “components” that contain both.

There are obviously still cases we would want to separate HTML and js, but it would be usually backed by the same arguments you provide when introducing any other intermediate abstractions -- for example:
1. we want to reuse the HTML with different control logic
2. we want to test the stateless HTML component alone

Reply:
Thanks for the info! That makes sense to me.

Reply:
Thank you both for the ideas! 

KFP UI currently organizes frontend code using component as unit. At the current state, separating HTML and js scripts will potentially mean that HTML can be reused by multiple js control logics. Therefore changing one HTML file will impact multiple js files, which is less preferred for keeping focus of concern contained. 

The benefit of component is that we have input parameters (props) and visualization (HTML) contained in the same file, and component is portable and reusable as single unit. 

Feel free to continue discussion on this thread. Also FYI, we can create GitHub issue to discuss best practice, and link them to this umbrella issue: https://github.com/kubeflow/pipelines/issues/5118.